File: 107 Binary Tree Level Order Traversal II.java
/* Add to Queue(Using LinkedList), poll and add the same level's nodes to the list, and add to result one level at a time.*/
class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        if(root == null) return res;
        Queue<TreeNode> levelQ = new LinkedList<>();
        levelQ.add(root);
        while(q.size() > 0) {
            List<Integer> level = new LinkedList<>();
            int size = q.size();
            for(int i = 0; i < size; i++) {
                TreeNode node = q.poll();
                level.add(node);
                if(node.left != null) q.add(node.left);
                if(node.right != null) q.add(node.right);
            }
            res.add(0, level);
        }
        return res;
    }
}
/* Visit by level 
class Solution {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        visit(root, res, 0);
        Collections.reverse(res);
        return res;        
    }
    public void visit(TreeNode node, List<List<Integer>> res, int level) {
        if(node == null)
            return;
        if(res.size() < level + 1)
            res.add(new ArrayList<Integer>());
        res.get(level).add(node.val);
        visit(node.left, res, level+1);
        visit(node.right, res, level+1);
    }
}
*/

File: 107 Binary Tree Level Order Traversal II.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution:
    def levelOrderBottom(self, root: 'TreeNode') -> 'List[List[int]]':
        if not root: return []        
        q = []
        res = []
        q.append(root)
        while q:
            llist = []
            lsize = len(q)
            for i in range(lsize):
                n = q.pop(0)
                llist.append(n.val)
                if n.left: q.append(n.left)
                if n.right: q.append(n.right)
            res.insert(0, llist)
        return res
''' 
# Recursive Solution
class Solution:
    def levelOrderBottom(self, root: 'TreeNode') -> 'List[List[int]]':
        res = []
        self.dfs(root, 0, res)
        return res
    def dfs(self, root, level, res):
        if root:
            if len(res) < level + 1:
                res.insert(0, [])
            res[-(level+1)].append(root.val)
            self.dfs(root.left, level+1, res)
            self.dfs(root.right, level+1, res)
'''